src/lib/SaferNames/Name.hs

```diff
@@ -4,267 +4,113 @@
-  where freshName = freshRawName "v" (LM.keysSet scope)
-
-freshRawName :: D.Tag -> S.Set RawName -> RawName
-freshRawName tag usedNames = D.Name D.GenName tag nextNum
-  where
-    nextNum = case S.lookupLT (D.Name D.GenName tag bigInt) usedNames of
-                Just (D.Name D.GenName tag' i)
-                  | tag' /= tag -> 0
-                  | i < bigInt  -> i + 1
-                  | otherwise   -> error "Ran out of numbers!"
-                _ -> 0
-    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
-
-unitName :: Name UnitScope
-unitName = UnsafeMakeName "TheName"
-
-unitScope :: Scope UnitScope
-unitScope = UnsafeMakeEnv $ LM.singleton unitRawName ()
-  where UnsafeMakeName unitRawName = unitName
-
-envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
-envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
-  where f' rawName = f $ UnsafeMakeName rawName
-
-envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
-envPairs (UnsafeMakeEnv m) = do
-  let (names, vals) = unzip $ LM.assocs m
-  (unsafeMakePlainBinderList names, vals)
-
-unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
-unsafeMakePlainBinderList names = case names of
-  [] -> unsafeCoerceB Empty
-  (name:rest) -> Nest (UnsafeMakeBinder name) $ unsafeMakePlainBinderList rest
-
--- === specialized traversals ===
-
-type NameSet n = S.Set (Name n)
-
-fmapNames :: HasNamesE e => Scope b -> (Name a -> Name b) -> e a -> e b
-fmapNames scope f e = runIdentity $ traverseNamesE scope t e
-  where t = newNameTraversal (pure . f)
-
-foldMapNames :: (HasNamesE e, Monoid a) => (Name n -> a) -> e n -> a
-foldMapNames f e =
-  execWriter $ traverseNamesE unitScope t e
-  where t = newNameTraversal \name -> tell (f name) >> return unitName
-
-freeNames :: HasNamesE e => e n -> NameSet n
-freeNames e = foldMapNames S.singleton e
-
--- should be equivalent to `\ext -> fmapNames (injectName ext)`
-injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
-injectNamesL _ = unsafeCoerceE
-
--- TODO: consider returning one of the `Name (n:=>:l)` names found in the
--- `Nothing` case, or even the full set.
-projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
-projectNamesL scope scopeFragment e = traverseNamesE scope t e
-  where t = newNameTraversal \name -> case projectName scopeFragment name of
-                                        Left name' -> Just name'
-                                        Right _    -> Nothing
+(<>>) :: Env v i o -> EnvFrag v i i' o -> Env v i' o
+(<>>) (Env f frag) (EnvFrag frag') = Env f $ concatNameMaps frag frag'
+
+infixl 1 <.>
+(<.>) :: EnvFrag v i1 i2 o -> EnvFrag v i2 i3 o -> EnvFrag v i1 i3 o
+(<.>) (EnvFrag env) (EnvFrag env') = EnvFrag (concatNameMaps env env')
+
+emptyEnvFrag :: EnvFrag v i i o
+emptyEnvFrag = EnvFrag emptyNameMap
+
+lookupEnvFrag :: EnvFrag v i i' o -> Name s i' -> Either (Name s i) (v s o)
```

Comment:
understanding check: we either prove that this name wasn't in the fragment (`Left name'`) or we have a value from the fragment `Right (lookupNameMap m name) :: v s o`

Reply:
yep, exactly